### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by ismaelherrera1**
Module 3 functionalities handled:
Navigation between pages,
Json consumption from local file,
ListView widgets implemented.
[pr_module3_video.zip](https://github.com/ismaelherrera1/TheMovieDB/files/12524863/pr_module3_video.zip)

### CODE REVIEW

File: lib/interfaces/i_movie_repository.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import 'dart:convert';
+import 'package:flutter/services.dart';
+import '../models/genre.dart';
+import '../models/movie.dart';
+
+abstract class IRepository<T> {
+  Future<List<T>> readData();
```
Comment (matiasilveiro-gl) [inline lib/interfaces/i_movie_repository.dart]: I'd mark this method to return `Future<T>`, to leave it as generic as possible

File: lib/interfaces/i_movie_repository.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+class MovieRepository extends IRepository {
+  List<Movie> movies = [];
```
Comment (matiasilveiro-gl) [inline lib/interfaces/i_movie_repository.dart]: Unused variable

File: lib/interfaces/i_movie_repository.dart [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  Future<List<Movie>> readData() async {
+    const String path = 'assets/mocks/movies.json';
+    final String jsonString = await rootBundle.loadString(path);
+    List<dynamic> movieData = jsonDecode(jsonString);
+    return movieData.map((e) => Movie.fromJson(e)).toList();
+  }
+}
+
+class GenreRepository extends IRepository {
+  List<Genre> genres = [];
```
Comment (matiasilveiro-gl) [inline lib/interfaces/i_movie_repository.dart]: Unused variable

File: lib/interfaces/i_movie_repository.dart
Comment (matiasilveiro-gl) [inline lib/interfaces/i_movie_repository.dart:1]: Please split interface and implementations in different files

File: lib/screens/home_page.dart [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  @override
+  State<HomePage> createState() => _HomePageState();
+}
+
+class _HomePageState extends State<HomePage> {
+  IRepository moviesRepo = MovieRepository();
+  IRepository genreRepo = GenreRepository();
+  late Future<List<dynamic>> movies;
+  late Future<List<dynamic>> genres;
```
Comment (matiasilveiro-gl) [inline lib/screens/home_page.dart]: Avoid using `dynamic` in this case. You know the return type, make use of the type cast!

File: lib/screens/home_page.dart [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                      itemCount: snapshot.data!.length,
+                      itemBuilder: (BuildContext context, int index) {
+                        return Padding(
+                          padding:
+                              const EdgeInsets.all(UIConstants.defaultPadding),
+                          child: GenreTile(
+                            genre: snapshot.data![index],
+                          ),
+                        );
+                      });
```
Comment (matiasilveiro-gl) [inline lib/screens/home_page.dart]: Trailing comma

File: lib/utils/ui_constants.dart [hunk 6]
```diff
@@ -1,10 +1,17 @@
...
   static const double appBarRadiusBorder = 25;
   static const double defaultPadding = 8.0;
-}
+  static const double genreScrollHeight = 40;
+  static const double genresBorderRadius = 5;
+  static const double genresShadowSpread = 1;
+  static const double genresBlurRadius = 7;
+  static const double genresOffsetDX = 2;
+  static const double genresOffsetDY = 4;
+}
```
Comment (matiasilveiro-gl) [inline lib/utils/ui_constants.dart]: EOF

File: lib/widgets/genre_tile.dart [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        shape: BoxShape.rectangle,
+        borderRadius: BorderRadius.circular(UIConstants.genresBorderRadius),
+        color: Colors.black,
+        boxShadow: const [
+          BoxShadow(
+            color: Colors.black87,
+            spreadRadius: UIConstants.genresShadowSpread,
+            blurRadius: UIConstants.genresBlurRadius,
+            offset:
+                Offset(UIConstants.genresOffsetDX, UIConstants.genresOffsetDY),
```
Comment (matiasilveiro-gl) [inline lib/widgets/genre_tile.dart]: Trailing comma

File: lib/widgets/movie_tile.dart [hunk 8]
```diff
@@ -0,0 +1,66 @@
+import 'package:flutter/material.dart';
+import 'package:the_movie_db/screens/movie_details.dart';
```
Comment (matiasilveiro-gl) [inline lib/widgets/movie_tile.dart]: Relative import

File: lib/widgets/movie_tile.dart [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      children: [
+        Expanded(
+          child: Container(
+            padding: const EdgeInsets.all(UIConstants.defaultPadding),
+            height: 200,
+            alignment: Alignment.bottomLeft,
+            decoration: BoxDecoration(
+              image: DecorationImage(
+                  image: NetworkImage(movie.fullBackdropPath),
+                  fit: BoxFit.cover),
```
Comment (matiasilveiro-gl) [inline lib/widgets/movie_tile.dart]: Trailing comma

File: lib/widgets/nav_bar.dart [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          BottomNavigationBarItem(
+            icon: Icon(
+              Icons.person_2_outlined,
+            ),
+            activeIcon: Icon(
+              Icons.person,
+            ),
+            label: 'Profile',
+          )
+        ]);
```
Comment (matiasilveiro-gl) [inline lib/widgets/nav_bar.dart]: Trailing comma

File: lib/interfaces/i_movie_repository.dart
Comment (ismaelherrera1) [reply lib/interfaces/i_movie_repository.dart:1]: 🏃

File: lib/interfaces/i_movie_repository.dart [hunk 11]
```diff
@@ -0,0 +1,6 @@
+
+abstract class IRepository<T> {
+  Future<T> readMovieData();
+  Future<T> readGenreData();
```
Comment (matiasilveiro-gl) [inline lib/interfaces/i_movie_repository.dart]: In this case you don't need the generic `<T>` type, as both `readxxxData()` methods are pretty specific with their return type. Replace with the correct return type

**Review by matiasilveiro-gl — APPROVED** [review]
LGTM! With minor comment

File: lib/widgets/movie_tile.dart [hunk 12]
```diff
@@ -35,10 +35,12 @@ class MovieTile extends StatelessWidget {
               children: [
                 Text(
                   movie.title,
+                  overflow: TextOverflow.fade,
+                  softWrap: false,
+                  maxLines: 1,
```
Comment (Kevin-Guillen) [inline lib/widgets/movie_tile.dart]: constant

**Review by Kevin-Guillen — APPROVED** [review]

File: lib/models/movie.dart [hunk 13]
```diff
@@ -32,31 +32,22 @@ class Movie {
...
+      id: json["id"] as int,
+      originalLanguage: json["original_language"] as String,
+      overview: json["overview"] as String,
+      popularity: json["popularity"] as double,
+      posterPath: json["poster_path"] as String,
+      releaseDate: json["release_date"] as String,
+      title: json["title"] as String,
+      video: json["video"] as bool,
+      voteAverage: json["vote_average"].toDouble(),
+      voteCount: json["vote_count"] as int,
```
Comment (FranciscoGaraventa) [inline lib/models/movie.dart]: Why you are casting all the parameters?

File: lib/screens/home_page.dart [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: Column(
+        children: [
+          SizedBox(
+            height: UIConstants.genreScrollHeight,
+            child: FutureBuilder(
+              future: genres,
+              builder: (BuildContext context, AsyncSnapshot snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/screens/home_page.dart]: trailing comma

File: lib/screens/home_page.dart [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+                    },
+                  );
+                }
+              },
+            ),
+          ),
+          Expanded(
+            child: FutureBuilder(
+              future: movies,
+              builder: (BuildContext context, AsyncSnapshot snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/screens/home_page.dart]: trailing comma

File: lib/widgets/movie_overview.dart [hunk 16]
```diff
@@ -33,12 +32,30 @@ class MovieOverview extends StatelessWidget {
...
-                Text(
-                  '$genresSubtitle $genreIds',
-                  style: const TextStyle(
-                    fontSize: UIConstants.genresFontSize,
-                    fontFamily: 'Poppins',
-                  ),
+                FutureBuilder(
+                  future: genres,
+                  builder:
+                      (BuildContext context, AsyncSnapshot<dynamic> snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/widgets/movie_overview.dart]: trailing comma

File: lib/models/movie.dart (hunk 13 above)
Comment (ismaelherrera1) [reply lib/models/movie.dart]: didnt realize its pointless. This is from an earlier version, right now it does not state any error without the casting! ty

File: lib/screens/home_page.dart [hunk 17]
```diff
@@ -34,7 +34,7 @@ class _HomePageState extends State<HomePage> {
             height: UIConstants.genreScrollHeight,
             child: FutureBuilder(
               future: genres,
-              builder: (BuildContext context, AsyncSnapshot snapshot) {
+              builder: (BuildContext context, AsyncSnapshot snapshot,) {
```
Comment (Kevin-Guillen) [inline lib/screens/home_page.dart]: format code

File: lib/screens/home_page.dart [hunk 18]
```diff
@@ -68,7 +68,7 @@ class _HomePageState extends State<HomePage> {
           Expanded(
             child: FutureBuilder(
               future: movies,
-              builder: (BuildContext context, AsyncSnapshot snapshot) {
+              builder: (BuildContext context, AsyncSnapshot snapshot,) {
```
Comment (Kevin-Guillen) [inline lib/screens/home_page.dart]: format code

File: lib/widgets/movie_overview.dart [hunk 19]
```diff
@@ -35,7 +35,7 @@ class MovieOverview extends StatelessWidget {
                 FutureBuilder(
                   future: genres,
                   builder:
-                      (BuildContext context, AsyncSnapshot<dynamic> snapshot) {
+                      (BuildContext context, AsyncSnapshot<dynamic> snapshot,) {
```
Comment (Kevin-Guillen) [inline lib/widgets/movie_overview.dart]: format code

**Review by FranciscoGaraventa — APPROVED** [review]
LG

**Review by Kevin-Guillen — APPROVED** [review]
lgtm

**Pull request merged**